docs/abi-routing.md

```diff
@@ -0,0 +1,48 @@
+
+### Approval Program
+If the program runs without error and returns a non-zero result, the transaction is allowed (though another transaction in the group may cause the whole group to fail as transaction groups are all or nothing).
+
+### Clear State Program
+A transaction which invokes the clear state program will be allowed even if the logic errors.
+
+## Application creating
+
+The Application ID (`apid`) field determines if this is an application create call or not. If it is set to `0` then the call will create a new application using the program bytes attached to this transaction. If it is set to any other number, then the call will be processed by the application on the target network which has the matching id. If such an application does not exist, the transaction will fail. It is worth noting here that the same contract can be deployed to multiple networks, or multiple times to the same network; and it will receive a different application id.
+
+If a new application is created, the transaction confirmation response will contain the id of the application.
+
+## On Completion Actions
+
+[On completion actions](https://developer.algorand.org/docs/get-details/dapps/avm/teal/specification/#oncomplete) are used for various application lifecycle events. They can be best interpreted as "The action that will have taken place upon completion of the processing of that transaction" (though it varies between these actions whether they occur before or after running the associated program).
+
+## ABI method selector vs. Bare methods
+
+[ARC004](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md) describes, among other things, a standard for routing application call transactions to specific methods of a contract. The approach is to use the first argument from the transaction's app args array (`apaa`) as a method selector. A method selector is defined as the first 4 bytes of the sha-512/256 hash of the method's arc4 signature. For example the method `sayHello(name: string, age: uint64): string` would have an arc4 signature of `sayHello(string,uint64)string`. Essentially the parameter names are ignored and we concatenate the param types. The hash of this string is `6ce0f4b5012a671bbe1b8355fe8f09f4aaa3b950115db8c71ee26162ecee8b9e` and the method selector is made of the first 4 bytes of this sequence. ie. `6ce0f4b5`.
+
+When a program processes an application call it should check the value of the first application arg. If there are no application args, then this considered a `bare` request, otherwise the method which a matching method selector should be invoked and the remainder of the application args should be forwarded to that method.
```

Comment:
```suggestion
When a program processes an application call it should check the value of the first application arg. If there are no application args, then this considered a `bare` request, otherwise the method with a matching method selector should be invoked and the remainder of the application args should be forwarded to that method.
```